noodles-editor/src/noodles/noodles.tsx

```diff
@@ -139,48 +149,90 @@ function useTheatreJs(projectName?: string) {
     // objects since we're storing that state in Theatre
     const sheetsById = Object.fromEntries(
       Object.entries(
-        timeline.sheetsById as Record<string, { staticOverrides?: { byObject?: unknown } }>
+        timeline.sheetsById as Record<
+          string,
+          { staticOverrides?: { byObject?: unknown } }
+        >,
       ).map(([sheetId, sheet]) => [
         sheetId,
         {
           ...sheet,
           staticOverrides: {
-            byObject: pick(sheet.staticOverrides?.byObject || {}, ['editor', 'render']),
+            byObject: pick(sheet.staticOverrides?.byObject || {}, [
+              'editor',
+              'render',
+            ]),
           },
         },
-      ])
+      ]),
     )
 
     return { ...timeline, sheetsById }
   }, [theatreState.name])
 
-  return { theatreReady, theatreProject, theatreSheet, setTheatreProject, getTimelineJson }
+  return {
+    theatreReady,
+    theatreProject,
+    theatreSheet,
+    setTheatreProject,
+    getTimelineJson,
+  }
 }
 
 // Not using the top-level sheet since a Noodles theatre sheet and project are dynamically created.
 // Also, the top-level sheet is used for theatre-managed project files, whereas a Noodles project file is managed within this visType.
+
 export function getNoodles(): Visualization {
-  const [projectName, setProjectName] = useState<string>()
+  const [, navigate] = useLocation()
+  const params = useParams()
+
+  // Get projectId from route params (/examples/:projectId) - router is single source of truth
+  const projectName = params.projectId
+
   const [showProjectNotFoundDialog, setShowProjectNotFoundDialog] = useState(false)
   const storageType = useActiveStorageType()
   const { setCurrentDirectory, setActiveStorageType, setError } = useFileSystemStore()
   const { theatreReady, theatreProject, theatreSheet, setTheatreProject, getTimelineJson } =
     useTheatreJs(projectName)
   const [nodes, setNodes, onNodesChange] = useNodesState<AnyNodeJSON>([])
-  const [edges, setEdges, onEdgesChange] = useEdgesState<ReactFlowEdge<unknown>>([])
+  const [edges, setEdges, onEdgesChange] = useEdgesState<
+    ReactFlowEdge<unknown>
+  >([])
   const vPressed = useKeyPress('v')
   const aPressed = useKeyPress('a')
   const [showChatPanel, setShowChatPanel] = useState(false)
 
+  // Update URL when project name changes (for when loading a project from file/storage)
+  // This updates the router, which will trigger projectName to update via useMemo
+  const setProjectName = useCallback(
+    (nameOrUpdater: React.SetStateAction<string | null>) => {
+      // Handle both direct values and updater functions
+      const name =
+        typeof nameOrUpdater === 'function'
+          ? nameOrUpdater(projectName ?? null)
```

Comment:
What's an updater function?

Reply:
`setState` can take either a value (`setVal(5)`) or a callback that is passed the previous value (`setVal(v => v + 1)`). See the SetStateAction type and the [docs](https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state). We would never really use the callback form but I guess since it's wrapping the `setProjectName` the AI is being overly cautious

Reply:
Gotcha, seems legacy though since this is a callback now rather than a true react state. Can the robot just update the types on component interfaces and simplify, or is there some 3rd party type using set state handle?